File: frontend/src/types/buildingScene.ts
// TypeScript contract for the stacked 3D building viewer (subfeature B).
// Mirrors docs/features/stacked-3d-viewer/05-api-contract.md exactly.

// Metric, Three.js-ready transform applied frontend-side as group.scale /
// group.rotation.y / group.position (Y-up, X-Z ground). `ty` equals `elevation_m`.
export interface ScenePlacement {
  scale: number;
  rotation_y_rad: number;
  tx: number;
  ty: number;
  tz: number;
}

export interface SceneFloor {
  floor_id: number;
  number: number;
  elevation_m: number;
  has_mesh: boolean;
  // ready GLB URL, or null when the floor has no assembled mesh
  mesh_url: string | null;
  // null when the floor is unsolved/unlinked (skipped in the scene); the
  // reference floor always gets an identity placement
  placement: ScenePlacement | null;
}

export interface BuildingScene3DResponse {
  building_id: number;
  // lowest floor (the world-frame origin); null when the building has no floors
  reference_floor_id: number | null;
  // FLOOR_HEIGHT echoed so the frontend never hardcodes it
  floor_height_m: number;
  // ordered by `number` ASC
  floors: SceneFloor[];
}

File: backend/app/models/building_scene.py
"""Pydantic models for the stacked 3D building viewer (subfeature B).

Exact JSON shapes are the contract in
``docs/features/stacked-3d-viewer/05-api-contract.md``. These models ARE that contract —
the ``GET /buildings/{id}/scene-3d`` response serialises through them. Pure declarations
only; no business logic.

``ScenePlacement`` is the metric, Three.js-ready transform the frontend applies as
``group.scale`` / ``group.rotation.y`` / ``group.position``. It is derived from
subfeature A's ``building_transform`` by the PURE ``processing.building_stack.floor_placement``
(see 06-pipeline-spec) — these models never recompute it.
"""

from typing import Optional

from pydantic import BaseModel


class ScenePlacement(BaseModel):
    """A floor GLB's placement in the building (reference-floor) metric world frame.

    Applied frontend-side as ``group.scale = scale``;
    ``group.rotation.y = rotation_y_rad``; ``group.position = (tx, ty, tz)``. Metres,
    Three.js Y-up with the ground in the X-Z plane. ``ty`` equals the floor's
    ``elevation_m``.
    """

    scale: float
    rotation_y_rad: float
    tx: float
    ty: float
    tz: float


class SceneFloor(BaseModel):
    """One floor in the 3D scene.

    ``has_mesh`` / ``mesh_url`` reflect ``Floor.mesh_file_glb`` (the assembled floor GLB);
    ``placement`` is ``None`` for an unsolved/unlinked floor (skipped in the scene, labelled
    in the side list). The reference (lowest) floor always gets an identity placement.
    """

    floor_id: int
    number: int
    elevation_m: float
    has_mesh: bool
    mesh_url: Optional[str] = None
    placement: Optional[ScenePlacement] = None


class BuildingScene3DResponse(BaseModel):
    """GET /buildings/{building_id}/scene-3d — the stacked 3D scene payload.

    ``reference_floor_id`` is the lowest floor (the world-frame origin), ``None`` when the
    building has no floors. ``floor_height_m`` echoes ``FLOOR_HEIGHT`` so the frontend never
    hardcodes it. ``floors`` is ordered by ``number`` ASC.
    """

    building_id: int
    reference_floor_id: Optional[int] = None
    floor_height_m: float
    floors: list[SceneFloor]
